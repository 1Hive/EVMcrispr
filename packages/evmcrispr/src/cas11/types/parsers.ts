import type { Parser } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

import type { Node } from './ast';

export type NodeParser<T = Node> = Parser<T, string, NodeParserState>;
