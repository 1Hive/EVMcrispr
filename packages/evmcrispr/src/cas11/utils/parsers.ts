import type { Err } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

export const buildParserError = (
  { index, error }: Err<string, null> | Err<string, NodeParserState>,
  type: string,
  msg?: string,
): string => {
  const splitRes = error.split('got');
  const wrongValueEncountered =
    splitRes.length === 2 ? splitRes[1].trim() : null;

  const errorMessage = msg
    ? `${msg}${wrongValueEncountered ? `, got ${wrongValueEncountered}` : ''}`
    : error.split('): ')[1];
  return `${type}(col: ${index}): ${errorMessage}`;
};

export const getIncorrectReceivedValue = (errorMsg: string): string => {
  const splitRes = errorMsg.split('got ');

  if (splitRes.length === 2) {
    return `, got ${splitRes[1].trim()}`;
  }

  return '';
};
