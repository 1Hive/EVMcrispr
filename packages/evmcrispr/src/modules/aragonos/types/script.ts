import type { Address } from "viem";

/**
 * A call script.
 */
export interface CallScriptAction {
  /**
   * The action's target.
   */
  to: Address;
  /**
   * The action's calldata.
   */
  data: `0x${string}`;
}
