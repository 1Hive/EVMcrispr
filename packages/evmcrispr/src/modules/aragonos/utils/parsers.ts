import type { PublicClient } from "viem";
import { parseAbiItem } from "viem";

import type { Address } from "../../../types";
import type { ParsedApp, Repo } from "../types";
import { getSystemApp, isSystemApp } from "./interfaces";

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return "";
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : "";
};

const fetchImplementationAddress = (
  appAddress: Address,
  client: PublicClient,
): Promise<Address> => {
  return client.readContract({
    address: appAddress,
    abi: [
      parseAbiItem("function implementation() public view returns (address)"),
    ],
    functionName: "implementation",
  });
};

export const parseApp = async (
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  app: any,
  client: PublicClient,
): Promise<ParsedApp> => {
  const { address, appId, roles, version } = app;
  const { name: repoName, registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress:
      codeAddress ?? (await fetchImplementationAddress(address, client)),
    contentUri,
    name,
    registryName: registry?.name,
    roles: (roles as any[]).map((role) => ({
      ...role,
      roleHash: role["hash"] ?? role["roleHash"],
    })),
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};
