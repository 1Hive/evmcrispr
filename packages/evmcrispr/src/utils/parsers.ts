import { getSystemApp, isSystemApp } from '.';
import type { ParsedApp, Repo } from '../types';

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return '';
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : '';
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseApp = (app: any): ParsedApp => {
  const { address, appId, implementation, repoName, roles, version } = app;
  const { registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress: codeAddress ?? implementation.address,
    contentUri,
    name,
    registryName: registry?.name,
    roles,
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};

export const timeUnits: { [key: string]: number } = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  mo: 2592000,
  y: 31536000,
};
