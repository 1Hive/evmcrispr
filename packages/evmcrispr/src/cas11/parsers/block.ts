import type { Parser } from 'arcsecond';
import { char, coroutine, recursiveParser } from 'arcsecond';

import type { BlockExpressionNode } from '../types';
import { NodeType } from '../types';
import { commandExpressionParser } from './command';
import {
  endOfLine,
  optionalEmptyLines,
  optionalWhitespace,
  surroundedBy,
} from './utils';

export const blockExpressionParser: Parser<BlockExpressionNode, string, any> =
  recursiveParser(() =>
    coroutine(function* () {
      yield char('(');

      yield optionalWhitespace;

      yield endOfLine;

      const scopedCommands = yield optionalEmptyLines(commandExpressionParser);

      yield surroundedBy(optionalWhitespace)(char(')'));

      return {
        type: NodeType.BlockExpression,
        body: scopedCommands,
      } as unknown as BlockExpressionNode;
    }),
  );
