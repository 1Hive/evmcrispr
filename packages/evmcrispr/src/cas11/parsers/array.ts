import type { Parser } from 'arcsecond';
import {
  between,
  char,
  coroutine,
  recursiveParser,
  sequenceOf,
} from 'arcsecond';

import type { ArrayExpressionNode } from '../types';
import { NodeType } from '../types';

import { argumentExpressionParser } from './expression';

import { commaSeparated, optionalWhitespace } from './utils';

export const arrayExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    between(sequenceOf([char('['), optionalWhitespace]))(
      sequenceOf([optionalWhitespace, char(']')]),
    )(commaSeparated(argumentExpressionParser)).map(
      (elements: any): ArrayExpressionNode => {
        return {
          type: NodeType.ArrayExpression,
          elements,
        };
      },
    ),
);

export const arrayExpresionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('[');
    yield optionalWhitespace;
    const elements = yield commaSeparated(argumentExpressionParser);
    yield optionalWhitespace;
    yield char(']');

    return {
      type: NodeType.ArrayExpression,
      elements,
    };
  }),
);
