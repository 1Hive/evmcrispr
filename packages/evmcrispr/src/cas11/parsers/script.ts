import type { Parser } from 'arcsecond';
import { coroutine, toValue } from 'arcsecond';
import type { AST } from 'prettier';

import { deepConsoleLog } from '../../../test/test-helpers/cas11';

import { ASTType } from '../types';
import { commandExpressionParser } from './command';
import { optionalEmptyLines } from './utils';

export const scriptParser: Parser<AST> = coroutine(function* () {
  const lines = yield optionalEmptyLines(commandExpressionParser);
  return {
    type: ASTType.Program,
    body: lines,
  };
});

export const parseScript = (script: string): AST | undefined => {
  const res = scriptParser.run(script);

  deepConsoleLog(res);
  if (res.isError) {
    console.log(res.error);
    return;
  }

  return toValue(res);
};
