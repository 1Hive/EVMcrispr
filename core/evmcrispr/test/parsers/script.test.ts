import { runParser } from '@1hive/evmcrispr-test-common';

import { scriptParser } from '../../src/parsers/script';

describe('Parsers - script', () => {
  it('should parse an script correctly', () => {
    const script = `
      load aragonos as ar
      load superfluid as sf\r\n
      
      ar:connect my-dao-ens (   
        forward token-manager voting      (
          install wrapper-hooked-token-manager.open 0x83E57888cd55C3ea1cfbf0114C963564d81e318d false 0
        
        
        )     




        forward token-manager voting agent (
          
          set $agent finance::vault()

          forward wrappable-token-manager.open disputable-voting.open agent (
            set $daix @token("fDAIx")







            
            sf:token approve @token('DAI') @me 15.45e18


            sf:batchcall (
              token upgrade $daix 4500.43e18
              flow create $daix $agent 1e18mo
              token downgrade @token('USDCx')
            )


          )
          
          
        ) 
        
        
                          )
  
  
      
      `;
    const parsedScript = runParser(scriptParser, script);

    expect(parsedScript).to.matchSnapshot();
  });
});
