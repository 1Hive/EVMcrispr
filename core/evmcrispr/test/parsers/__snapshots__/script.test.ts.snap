// Vitest Snapshot v1

exports[`Parsers - script > should parse an script correctly 1`] = `
Cas11AST {
  "body": [
    {
      "args": [
        {
          "left": {
            "loc": {
              "end": {
                "col": 19,
                "line": 2,
              },
              "start": {
                "col": 11,
                "line": 2,
              },
            },
            "type": "ProbableIdentifier",
            "value": "aragonos",
          },
          "loc": {
            "end": {
              "col": 25,
              "line": 2,
            },
            "start": {
              "col": 11,
              "line": 2,
            },
          },
          "right": {
            "loc": {
              "end": {
                "col": 25,
                "line": 2,
              },
              "start": {
                "col": 23,
                "line": 2,
              },
            },
            "type": "ProbableIdentifier",
            "value": "ar",
          },
          "type": "AsExpression",
        },
      ],
      "loc": {
        "end": {
          "col": 25,
          "line": 2,
        },
        "start": {
          "col": 6,
          "line": 2,
        },
      },
      "name": "load",
      "opts": [],
      "type": "CommandExpression",
    },
    {
      "args": [
        {
          "left": {
            "loc": {
              "end": {
                "col": 21,
                "line": 3,
              },
              "start": {
                "col": 11,
                "line": 3,
              },
            },
            "type": "ProbableIdentifier",
            "value": "superfluid",
          },
          "loc": {
            "end": {
              "col": 27,
              "line": 3,
            },
            "start": {
              "col": 11,
              "line": 3,
            },
          },
          "right": {
            "loc": {
              "end": {
                "col": 27,
                "line": 3,
              },
              "start": {
                "col": 25,
                "line": 3,
              },
            },
            "type": "ProbableIdentifier",
            "value": "sf",
          },
          "type": "AsExpression",
        },
      ],
      "loc": {
        "end": {
          "col": 27,
          "line": 3,
        },
        "start": {
          "col": 6,
          "line": 3,
        },
      },
      "name": "load",
      "opts": [],
      "type": "CommandExpression",
    },
    {
      "args": [
        {
          "loc": {
            "end": {
              "col": 27,
              "line": 6,
            },
            "start": {
              "col": 17,
              "line": 6,
            },
          },
          "type": "ProbableIdentifier",
          "value": "my-dao-ens",
        },
        {
          "body": [
            {
              "args": [
                {
                  "loc": {
                    "end": {
                      "col": 29,
                      "line": 7,
                    },
                    "start": {
                      "col": 16,
                      "line": 7,
                    },
                  },
                  "type": "ProbableIdentifier",
                  "value": "token-manager",
                },
                {
                  "loc": {
                    "end": {
                      "col": 36,
                      "line": 7,
                    },
                    "start": {
                      "col": 30,
                      "line": 7,
                    },
                  },
                  "type": "ProbableIdentifier",
                  "value": "voting",
                },
                {
                  "body": [
                    {
                      "args": [
                        {
                          "loc": {
                            "end": {
                              "col": 51,
                              "line": 8,
                            },
                            "start": {
                              "col": 18,
                              "line": 8,
                            },
                          },
                          "type": "ProbableIdentifier",
                          "value": "wrapper-hooked-token-manager.open",
                        },
                        {
                          "loc": {
                            "end": {
                              "col": 94,
                              "line": 8,
                            },
                            "start": {
                              "col": 52,
                              "line": 8,
                            },
                          },
                          "type": "AddressLiteral",
                          "value": "0x83E57888cd55C3ea1cfbf0114C963564d81e318d",
                        },
                        {
                          "loc": {
                            "end": {
                              "col": 100,
                              "line": 8,
                            },
                            "start": {
                              "col": 95,
                              "line": 8,
                            },
                          },
                          "type": "BoolLiteral",
                          "value": false,
                        },
                        {
                          "loc": {
                            "end": {
                              "col": 102,
                              "line": 8,
                            },
                            "start": {
                              "col": 101,
                              "line": 8,
                            },
                          },
                          "type": "NumberLiteral",
                          "value": "0",
                        },
                      ],
                      "loc": {
                        "end": {
                          "col": 102,
                          "line": 8,
                        },
                        "start": {
                          "col": 10,
                          "line": 8,
                        },
                      },
                      "name": "install",
                      "opts": [],
                      "type": "CommandExpression",
                    },
                  ],
                  "loc": {
                    "end": {
                      "col": 9,
                      "line": 11,
                    },
                    "start": {
                      "col": 42,
                      "line": 7,
                    },
                  },
                  "type": "BlockExpression",
                },
              ],
              "loc": {
                "end": {
                  "col": 9,
                  "line": 11,
                },
                "start": {
                  "col": 8,
                  "line": 7,
                },
              },
              "name": "forward",
              "opts": [],
              "type": "CommandExpression",
            },
            {
              "args": [
                {
                  "loc": {
                    "end": {
                      "col": 29,
                      "line": 16,
                    },
                    "start": {
                      "col": 16,
                      "line": 16,
                    },
                  },
                  "type": "ProbableIdentifier",
                  "value": "token-manager",
                },
                {
                  "loc": {
                    "end": {
                      "col": 36,
                      "line": 16,
                    },
                    "start": {
                      "col": 30,
                      "line": 16,
                    },
                  },
                  "type": "ProbableIdentifier",
                  "value": "voting",
                },
                {
                  "loc": {
                    "end": {
                      "col": 42,
                      "line": 16,
                    },
                    "start": {
                      "col": 37,
                      "line": 16,
                    },
                  },
                  "type": "ProbableIdentifier",
                  "value": "agent",
                },
                {
                  "body": [
                    {
                      "args": [
                        {
                          "loc": {
                            "end": {
                              "col": 20,
                              "line": 18,
                            },
                            "start": {
                              "col": 14,
                              "line": 18,
                            },
                          },
                          "type": "VariableIdentifier",
                          "value": "$agent",
                        },
                        {
                          "args": [],
                          "loc": {
                            "end": {
                              "col": 37,
                              "line": 18,
                            },
                            "start": {
                              "col": 21,
                              "line": 18,
                            },
                          },
                          "method": "vault",
                          "target": {
                            "loc": {
                              "end": {
                                "col": 28,
                                "line": 18,
                              },
                              "start": {
                                "col": 21,
                                "line": 18,
                              },
                            },
                            "type": "ProbableIdentifier",
                            "value": "finance",
                          },
                          "type": "CallExpression",
                        },
                      ],
                      "loc": {
                        "end": {
                          "col": 37,
                          "line": 18,
                        },
                        "start": {
                          "col": 10,
                          "line": 18,
                        },
                      },
                      "name": "set",
                      "opts": [],
                      "type": "CommandExpression",
                    },
                    {
                      "args": [
                        {
                          "loc": {
                            "end": {
                              "col": 46,
                              "line": 20,
                            },
                            "start": {
                              "col": 18,
                              "line": 20,
                            },
                          },
                          "type": "ProbableIdentifier",
                          "value": "wrappable-token-manager.open",
                        },
                        {
                          "loc": {
                            "end": {
                              "col": 69,
                              "line": 20,
                            },
                            "start": {
                              "col": 47,
                              "line": 20,
                            },
                          },
                          "type": "ProbableIdentifier",
                          "value": "disputable-voting.open",
                        },
                        {
                          "loc": {
                            "end": {
                              "col": 75,
                              "line": 20,
                            },
                            "start": {
                              "col": 70,
                              "line": 20,
                            },
                          },
                          "type": "ProbableIdentifier",
                          "value": "agent",
                        },
                        {
                          "body": [
                            {
                              "args": [
                                {
                                  "loc": {
                                    "end": {
                                      "col": 21,
                                      "line": 21,
                                    },
                                    "start": {
                                      "col": 16,
                                      "line": 21,
                                    },
                                  },
                                  "type": "VariableIdentifier",
                                  "value": "$daix",
                                },
                                {
                                  "args": [
                                    {
                                      "loc": {
                                        "end": {
                                          "col": 36,
                                          "line": 21,
                                        },
                                        "start": {
                                          "col": 29,
                                          "line": 21,
                                        },
                                      },
                                      "type": "StringLiteral",
                                      "value": "fDAIx",
                                    },
                                  ],
                                  "loc": {
                                    "end": {
                                      "col": 37,
                                      "line": 21,
                                    },
                                    "start": {
                                      "col": 22,
                                      "line": 21,
                                    },
                                  },
                                  "name": "token",
                                  "type": "HelperFunctionExpression",
                                },
                              ],
                              "loc": {
                                "end": {
                                  "col": 37,
                                  "line": 21,
                                },
                                "start": {
                                  "col": 12,
                                  "line": 21,
                                },
                              },
                              "name": "set",
                              "opts": [],
                              "type": "CommandExpression",
                            },
                            {
                              "args": [
                                {
                                  "loc": {
                                    "end": {
                                      "col": 28,
                                      "line": 30,
                                    },
                                    "start": {
                                      "col": 21,
                                      "line": 30,
                                    },
                                  },
                                  "type": "ProbableIdentifier",
                                  "value": "approve",
                                },
                                {
                                  "args": [
                                    {
                                      "loc": {
                                        "end": {
                                          "col": 41,
                                          "line": 30,
                                        },
                                        "start": {
                                          "col": 36,
                                          "line": 30,
                                        },
                                      },
                                      "type": "StringLiteral",
                                      "value": "DAI",
                                    },
                                  ],
                                  "loc": {
                                    "end": {
                                      "col": 42,
                                      "line": 30,
                                    },
                                    "start": {
                                      "col": 29,
                                      "line": 30,
                                    },
                                  },
                                  "name": "token",
                                  "type": "HelperFunctionExpression",
                                },
                                {
                                  "args": [],
                                  "loc": {
                                    "end": {
                                      "col": 46,
                                      "line": 30,
                                    },
                                    "start": {
                                      "col": 43,
                                      "line": 30,
                                    },
                                  },
                                  "name": "me",
                                  "type": "HelperFunctionExpression",
                                },
                                {
                                  "loc": {
                                    "end": {
                                      "col": 55,
                                      "line": 30,
                                    },
                                    "start": {
                                      "col": 47,
                                      "line": 30,
                                    },
                                  },
                                  "power": 18,
                                  "type": "NumberLiteral",
                                  "value": "15.45",
                                },
                              ],
                              "loc": {
                                "end": {
                                  "col": 55,
                                  "line": 30,
                                },
                                "start": {
                                  "col": 12,
                                  "line": 30,
                                },
                              },
                              "module": "sf",
                              "name": "token",
                              "opts": [],
                              "type": "CommandExpression",
                            },
                            {
                              "args": [
                                {
                                  "body": [
                                    {
                                      "args": [
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 27,
                                              "line": 34,
                                            },
                                            "start": {
                                              "col": 20,
                                              "line": 34,
                                            },
                                          },
                                          "type": "ProbableIdentifier",
                                          "value": "upgrade",
                                        },
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 33,
                                              "line": 34,
                                            },
                                            "start": {
                                              "col": 28,
                                              "line": 34,
                                            },
                                          },
                                          "type": "VariableIdentifier",
                                          "value": "$daix",
                                        },
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 44,
                                              "line": 34,
                                            },
                                            "start": {
                                              "col": 34,
                                              "line": 34,
                                            },
                                          },
                                          "power": 18,
                                          "type": "NumberLiteral",
                                          "value": "4500.43",
                                        },
                                      ],
                                      "loc": {
                                        "end": {
                                          "col": 44,
                                          "line": 34,
                                        },
                                        "start": {
                                          "col": 14,
                                          "line": 34,
                                        },
                                      },
                                      "name": "token",
                                      "opts": [],
                                      "type": "CommandExpression",
                                    },
                                    {
                                      "args": [
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 25,
                                              "line": 35,
                                            },
                                            "start": {
                                              "col": 19,
                                              "line": 35,
                                            },
                                          },
                                          "type": "ProbableIdentifier",
                                          "value": "create",
                                        },
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 31,
                                              "line": 35,
                                            },
                                            "start": {
                                              "col": 26,
                                              "line": 35,
                                            },
                                          },
                                          "type": "VariableIdentifier",
                                          "value": "$daix",
                                        },
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 38,
                                              "line": 35,
                                            },
                                            "start": {
                                              "col": 32,
                                              "line": 35,
                                            },
                                          },
                                          "type": "VariableIdentifier",
                                          "value": "$agent",
                                        },
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 45,
                                              "line": 35,
                                            },
                                            "start": {
                                              "col": 39,
                                              "line": 35,
                                            },
                                          },
                                          "power": 18,
                                          "timeUnit": "mo",
                                          "type": "NumberLiteral",
                                          "value": "1",
                                        },
                                      ],
                                      "loc": {
                                        "end": {
                                          "col": 45,
                                          "line": 35,
                                        },
                                        "start": {
                                          "col": 14,
                                          "line": 35,
                                        },
                                      },
                                      "name": "flow",
                                      "opts": [],
                                      "type": "CommandExpression",
                                    },
                                    {
                                      "args": [
                                        {
                                          "loc": {
                                            "end": {
                                              "col": 29,
                                              "line": 36,
                                            },
                                            "start": {
                                              "col": 20,
                                              "line": 36,
                                            },
                                          },
                                          "type": "ProbableIdentifier",
                                          "value": "downgrade",
                                        },
                                        {
                                          "args": [
                                            {
                                              "loc": {
                                                "end": {
                                                  "col": 44,
                                                  "line": 36,
                                                },
                                                "start": {
                                                  "col": 37,
                                                  "line": 36,
                                                },
                                              },
                                              "type": "StringLiteral",
                                              "value": "USDCx",
                                            },
                                          ],
                                          "loc": {
                                            "end": {
                                              "col": 45,
                                              "line": 36,
                                            },
                                            "start": {
                                              "col": 30,
                                              "line": 36,
                                            },
                                          },
                                          "name": "token",
                                          "type": "HelperFunctionExpression",
                                        },
                                      ],
                                      "loc": {
                                        "end": {
                                          "col": 45,
                                          "line": 36,
                                        },
                                        "start": {
                                          "col": 14,
                                          "line": 36,
                                        },
                                      },
                                      "name": "token",
                                      "opts": [],
                                      "type": "CommandExpression",
                                    },
                                  ],
                                  "loc": {
                                    "end": {
                                      "col": 13,
                                      "line": 37,
                                    },
                                    "start": {
                                      "col": 25,
                                      "line": 33,
                                    },
                                  },
                                  "type": "BlockExpression",
                                },
                              ],
                              "loc": {
                                "end": {
                                  "col": 13,
                                  "line": 37,
                                },
                                "start": {
                                  "col": 12,
                                  "line": 33,
                                },
                              },
                              "module": "sf",
                              "name": "batchcall",
                              "opts": [],
                              "type": "CommandExpression",
                            },
                          ],
                          "loc": {
                            "end": {
                              "col": 11,
                              "line": 40,
                            },
                            "start": {
                              "col": 76,
                              "line": 20,
                            },
                          },
                          "type": "BlockExpression",
                        },
                      ],
                      "loc": {
                        "end": {
                          "col": 11,
                          "line": 40,
                        },
                        "start": {
                          "col": 10,
                          "line": 20,
                        },
                      },
                      "name": "forward",
                      "opts": [],
                      "type": "CommandExpression",
                    },
                  ],
                  "loc": {
                    "end": {
                      "col": 9,
                      "line": 43,
                    },
                    "start": {
                      "col": 43,
                      "line": 16,
                    },
                  },
                  "type": "BlockExpression",
                },
              ],
              "loc": {
                "end": {
                  "col": 9,
                  "line": 43,
                },
                "start": {
                  "col": 8,
                  "line": 16,
                },
              },
              "name": "forward",
              "opts": [],
              "type": "CommandExpression",
            },
          ],
          "loc": {
            "end": {
              "col": 27,
              "line": 46,
            },
            "start": {
              "col": 28,
              "line": 6,
            },
          },
          "type": "BlockExpression",
        },
      ],
      "loc": {
        "end": {
          "col": 27,
          "line": 46,
        },
        "start": {
          "col": 6,
          "line": 6,
        },
      },
      "module": "ar",
      "name": "connect",
      "opts": [],
      "type": "CommandExpression",
    },
  ],
  "type": "Program",
}
`;
